### ISSUE TITLE
c.deletePositionsInList is broken

### ISSUE THREAD

**Issue by vitalije** [leo-editor/leo-editor#1539]
This method can delete some positions not in given list and also can skip deleting some positions in given list. Here is a script that shows both wrong behaviors
```python
pwbn = p
def appendnode(p, h):
    p1 = p.insertAsLastChild()
    p1.h = h
    return p1

def create_tree(p):
    while p.hasChildren():
        p.firstChild().doDelete()
    p1 = appendnode(p, 'a')
    appendnode(p, 'b')
    appendnode(p, 'c')
    # append clone of p1 
    p.v.children.append(p1.v)
    p1.v.parents.append(p.v)

create_tree(pwbn)
# children are a, b, c, <a clone>
g.es('created children:', [(x.h, x.childIndex()) for x in pwbn.children()])
# let's delete b and last a
aList = [pwbn.nthChild(1), pwbn.nthChild(3)]
g.es('deleting', [(x.h, x.childIndex()) for x in aList])
c.deletePositionsInList(aList)
# we expect that chlidren are a and c
g.es('rest of children', [(x.h, x.childIndex()) for x in pwbn.children()])
assert pwbn.nthChild(0).h == 'a', pwbn.nthChild(0).h
```

It creates three child nodes in the currently selected position, then appends a clone of the first child. After this selected node has four children with the headlines 'a', 'b', 'c' and 'a' (clone). Execute c.deletePositionsInList with the list containing positions of the second and fourth child. The expected result is that selected node has children 'a' and 'c'. The result is that node now contains three child nodes 'b', 'c' and 'a'. Nodes 'b' and last child 'a' were in the list but they are not deleted. The first node 'a' was not in the list and it has been deleted.

**Comment by vitalije** [comment]
I believe I have found the solution. Replacing the c.deletePositionsInList with the following:
```python
def deletePositionsInList(self, aList, callback=None, redraw=True):
    seen = set()
    for p in sorted(aList, reverse=True):
        par = p.stack[-1][0] if p.stack else c.hiddenRootNode
        x = par, p._childIndex
        if x not in seen:
            seen.add(x)
            p.doDelete()
    if redraw:
        c.redraw()
    return
```
My test script gives correct result and all unit test pass. In fact 918 ran, 3 failures, 2 errors and 65 skipped is the result both with this fix and without it.

Originally this method has a keyword argument `callback` which isn't used in Leo. It seems to me that it can be removed.

@edreamleo what do you think? Should I make these changes?

**Comment by edreamleo** [comment]
On Mon, Mar 16, 2020 at 1:44 PM Виталије Милошевић <notifications@github.com>
wrote:

> I believe I have found the solution. Replacing the c.deletePositionsInList
> with the following:
>
> def deletePositionsInList(self, aList, callback=None, redraw=True):
>     seen = set()
>     for p in sorted(aList, reverse=True):
>         par = p.stack[-1][0] if p.stack else c.hiddenRootNode
>         x = par, p._childIndex
>         if x not in seen:
>             seen.add(x)
>             p.doDelete()
>     if redraw:
>         c.redraw()
>     return
>
> My test script gives correct result and all unit test pass. In fact 918
> ran, 3 failures, 2 errors and 65 skipped is the result both with this fix
> and without it.
>
Iirc, handling nodes in reverse order doesn't guarantee that all will be
well. The culprits, as usual, are clones.

> Originally this method has a keyword argument callback which isn't used
> in Leo. It seems to me that it can be removed.
>
Yes, let's remove the callback kwarg.

> @edreamleo <https://github.com/edreamleo> what do you think? Should I
> make these changes?
>
Please make the changes in a separate branch, based on devel. I don't want
to change this for 6.2.

This is a tricky method, with a long history. Let's discuss this thoroughly
in 6.3.

Edward

**Comment by vitalije** [comment]
7bf61c368c00fe8 contains my implementation inside `c.deletePositionsInList`, disabled by default.

**Comment by vitalije** [comment]
Here is a test workbench. In the attached zip file there is a `test_outlines.zip` file which contains 50 randomly generated trees. All these trees have 30 unique nodes. About 6 of them are clonned.

There is also a test script `a.py`. This script accepts one argument an integer for selecting the outline from test_outlines.zip. It performs 1000 tests,
1. load selected outline
2. make a random list of positions
3. calculate dump value of the outline without selected positions
4. execute  `deletePositionsInList`
5. calculate dump value of the resulting outline
6. compare values (3) and (5) and reports error if they differ

If no error occurs in 1000 iterations, it prints dump value of the original outline.

After many experiments I haven't found any error.

[bug1539.zip](https://github.com/leo-editor/leo-editor/files/4345719/bug1539.zip)

**Comment by edreamleo** [comment]
On Tue, Mar 17, 2020 at 3:48 PM Виталије Милошевић <notifications@github.com>
wrote:

> Here is a test workbench. In the attached zip file there is a
> test_outlines.zip file which contains 50 randomly generated trees. All
> these trees have 30 unique nodes. About 6 of them are cloned.
>
Many thanks for this work.

When I awoke this morning, I realized that *you* are convinced the code is
sound. That's good enough for me. I'll incorporate your code in the 6.2 and
devel branches today.

A proper proof of correctness should not rely on the word 'obvious' :-)
I'll attempt that proof in a new thread.

Edward

**Comment by edreamleo** [comment]
I am going to close this issue. The new code has been merged into the 6.2 and devel branches.
